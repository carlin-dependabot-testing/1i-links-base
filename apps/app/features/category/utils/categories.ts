import { DEFAULT_CATEGORY } from '@/constants'
import { type Category } from '@/features/links/types/category.types'
import { getCategories } from '@/features/setup/utils/load-json-data'

/**
 * Returns a Set of category names in lowercase
 */
export const getCategoryNames = async (): Promise<Set<string>> => {
  const categories = await getCategories()

  return new Set([
    DEFAULT_CATEGORY,
    ...(categories?.map(cat => cat.name.toLowerCase()) || [])
  ])
}

/**
 * Returns all valid categories
 */
export const getValidCategories = async (): Promise<Category[]> => {
  const categories = await getCategories()
  return categories || []
}

/**
 * Validates if a category name exists
 */
export const isCategoryValid = async (
  categoryName: string
): Promise<boolean> => {
  const categoryNames = await getCategoryNames()
  return categoryNames.has(categoryName.toLowerCase())
}

/**
 * Returns a category by its slug
 */
export const getCategoryBySlug = async (slug: string) => {
  const categories = await getCategories()
  return categories?.find(category => category.name === slug) || null
}

// Synchronous versions with default values
export const getDefaultCategoryNames = (): Set<string> => {
  return new Set([DEFAULT_CATEGORY])
}

export const getDefaultCategories = (): Category[] => {
  return []
}
