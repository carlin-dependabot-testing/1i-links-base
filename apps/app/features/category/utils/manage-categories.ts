import fs from 'fs/promises'
import path from 'path'
import { z } from 'zod'

import { CategorySchema } from '@/features/links/schemas/category.schema'
import { type Category } from '@/features/links/types/category.types'

// Custom errors
export class CategoryError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'CategoryError'
  }
}

export const categoriesPath = path.join(
  process.cwd(),
  'public',
  'data',
  'categories.json'
)

/**
 * Retrieves all categories from the JSON file
 * @returns {Promise<Category[]>} Array of categories
 * @throws {CategoryError} If there's an error reading the file
 */
export async function getCategories(): Promise<Category[]> {
  try {
    const data = await fs.readFile(categoriesPath, 'utf8')

    if (data.trim().length === 0) {
      console.warn('The category file is empty')
      return []
    }

    try {
      const parsedData = JSON.parse(data)

      // Sort categories alphabetically by name (case-insensitive)
      return parsedData.sort((a: Category, b: Category) =>
        a.name.toLowerCase().localeCompare(b.name.toLowerCase())
      )
    } catch (parseError) {
      console.error('Error parsing JSON:', parseError)
      return []
    }
  } catch (error) {
    if (error instanceof z.ZodError) {
      console.error('Zod validation error:', error.errors)
      throw new CategoryError('Invalid data format in categories file')
    }
    console.error('Error reading category:', error)
    return []
  }
}

/**
 * Saves categories to the JSON file
 * @param {Category[]} category - Array of categories to save
 * @throws {CategoryError} If there's an error saving the file
 */
export async function saveCategories(category: Category[]): Promise<void> {
  try {
    const data = JSON.stringify(category, null, 2)
    await fs.writeFile(categoriesPath, data)
  } catch (error) {
    if (error instanceof z.ZodError) {
      throw new CategoryError('Invalid category data format')
    }
    throw new CategoryError('Failed to save categories')
  }
}

/**
 * Adds a new category to the storage
 * @param {Category} newCategory - Category to add
 * @returns {Promise<Category>} Added category
 * @throws {CategoryError} If the category is invalid or if there's an error saving
 */
export async function addCategory(category: Category): Promise<Category> {
  try {
    CategorySchema.parse(category)

    const categories = await getCategories()

    categories.push(category)
    await saveCategories(categories)

    return category
  } catch (error) {
    if (error instanceof z.ZodError) {
      throw new CategoryError('Invalid category data format')
    }
    throw new CategoryError('Failed to add category')
  }
}

export async function editCategory(
  oldName: string,
  newName: string,
  icon?: string,
  description?: string
): Promise<void> {
  try {
    if (!oldName?.trim() || !newName?.trim()) {
      throw new CategoryError('Invalid category name')
    }

    const categories = await getCategories()

    const categoryExists = categories.some(
      category => category.name === oldName
    )

    if (!categoryExists) {
      throw new CategoryError('Category not found')
    }

    const isDuplicate = categories.some(
      category => category.name === newName && category.name !== oldName
    )
    if (isDuplicate) {
      throw new CategoryError('A category with this name already exists')
    }

    const updatedCategories = categories.map((category: Category) => {
      if (category.name === oldName) {
        const updatedCategory = {
          ...category,
          name: newName,
          icon: icon || category.icon,
          description: description || category.description
        }

        // Validate the updated category
        CategorySchema.parse(updatedCategory)
        return updatedCategory
      }
      return category
    })

    await saveCategories(updatedCategories)
  } catch (error) {
    if (error instanceof z.ZodError) {
      throw new CategoryError('Invalid category data format')
    }
    throw new CategoryError('Failed to edit category')
  }
}

/**
 * Deletes a category from the storage
 * @param {string} name - Name of the category to delete
 * @throws {CategoryError} If the category name is invalid
 */
export async function deleteCategory(name: string): Promise<void> {
  try {
    if (typeof name !== 'string' || name.trim() === '') {
      throw new CategoryError('Invalid category name')
    }

    const categories = await getCategories()
    const updatedCategories = categories.filter(
      category => category.name !== name
    )
    await saveCategories(updatedCategories)
  } catch (error) {
    if (error instanceof CategoryError) {
      throw new CategoryError('Failed to delete category')
    }
    throw error
  }
}
