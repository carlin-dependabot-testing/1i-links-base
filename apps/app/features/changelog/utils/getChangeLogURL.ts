import appConfig from '@/config/app.config'

export function getChangeLogURL() {
  return `https://raw.githubusercontent.com/${appConfig.githubRepo}/refs/heads/main/CHANGELOG.md`
}
