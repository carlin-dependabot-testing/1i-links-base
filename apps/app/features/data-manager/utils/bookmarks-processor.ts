import { DEFAULT_CATEGORY, DEFAULT_COLOR } from '@/constants'
import { type LinksApp } from '@/features/links/types/link.types'

export async function processBookmarksFile(
  file: File
): Promise<Partial<LinksApp>[]> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader()

    reader.onload = event => {
      try {
        const text = event.target?.result as string
        const parser = new DOMParser()
        const doc = parser.parseFromString(text, 'text/html')
        const bookmarks = Array.from(doc.getElementsByTagName('a'))

        const links = bookmarks.map(bookmark => ({
          name: bookmark.textContent || '',
          environments: {
            production: bookmark.href
          },
          category: DEFAULT_CATEGORY,
          color: DEFAULT_COLOR
        }))

        resolve(links.filter(link => link.name && link.environments.production))
      } catch (error) {
        reject(new Error('Failed to process bookmarks file'))
      }
    }

    reader.onerror = () => {
      reject(new Error('Failed to read bookmarks file'))
    }

    reader.readAsText(file)
  })
}
