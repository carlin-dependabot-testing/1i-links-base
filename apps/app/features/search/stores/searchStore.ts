import { type StoreApi, createStore } from 'zustand'

export type SearchState = {
  searchTerm: string
}

type SearchActions = {
  setSearchTerm: (term: string) => void
}

export type SearchStore = SearchState & SearchActions

const defaultInitState: SearchState = {
  searchTerm: ''
}

export const createSearchStore = (
  initState: SearchState = defaultInitState
): StoreApi<SearchStore> => {
  return createStore<SearchStore>()(set => ({
    ...initState,
    setSearchTerm: (term: string) => set({ searchTerm: term })
  }))
}
