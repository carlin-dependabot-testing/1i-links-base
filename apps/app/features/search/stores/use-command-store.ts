import { create } from 'zustand'

interface CommandStore {
  isOpen: boolean
  setOpen: (open: boolean) => void
}

export const useCommandStore = create<CommandStore>(set => ({
  isOpen: false,
  setOpen: open => set({ isOpen: open })
}))
