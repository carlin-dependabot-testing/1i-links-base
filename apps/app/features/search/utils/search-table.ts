type SearchableObject = Record<string, any>

export function searchTable<T extends SearchableObject>(
  items: T[],
  searchTerm: string,
  searchFields: (keyof T)[]
): T[] {
  const normalizedTerm = searchTerm.toLowerCase().trim()

  if (!normalizedTerm) return items

  return items.filter(item =>
    searchFields.some(field => {
      const value = item[field]
      return String(value).toLowerCase().includes(normalizedTerm)
    })
  )
}
