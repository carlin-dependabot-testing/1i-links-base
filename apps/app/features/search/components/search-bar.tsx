'use client'

import { SearchIcon } from 'lucide-react'

import { useCommandStore } from '@/features/search/stores/use-command-store'

import { Button } from '@links-base/ui/button'
import { useIsMobile } from '@links-base/ui/hooks'

export const SearchBarButton = () => {
  return (
    <kbd
      className="pointer-events-none ml-auto inline-flex h-5 select-none items-center gap-1 rounded border bg-muted px-1.5 font-mono text-[10px] font-medium opacity-100"
      role="textbox"
      aria-label="keyboard shortcut"
    >
      <span className="text-xs">âŒ˜</span>K
    </kbd>
  )
}

export const SearchBar = () => {
  const { setOpen: setCommandOpen } = useCommandStore()
  const isMobile = useIsMobile()

  const handleClick = () => {
    setCommandOpen(true)
  }

  return (
    <div className="relative">
      <Button
        variant="outline"
        className={'w-full gap-2'}
        onClick={handleClick}
      >
        <SearchIcon className={'mr-2 h-4 w-4'} aria-label="search icon" />
        {!isMobile && <span>Search</span>}
        <SearchBarButton />
      </Button>
    </div>
  )
}
