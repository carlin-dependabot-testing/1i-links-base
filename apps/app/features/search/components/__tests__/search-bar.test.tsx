import { screen } from '@testing-library/react'
import { default as userEvent } from '@testing-library/user-event'

import { render } from '@/utils/test-utils'

import { useCommandStore } from '../../stores/use-command-store'
import { SearchBar } from '../search-bar'

// Mock the hooks
jest.mock('@links-base/ui/hooks', () => ({
  useIsMobile: () => false
}))

jest.mock('../../stores/use-command-store', () => ({
  useCommandStore: jest.fn() as unknown as typeof useCommandStore
}))

const mockSetOpen = jest.fn()

const renderComponent = (props = {}) => {
  return render(<SearchBar {...props} />)
}

describe('SearchBar', () => {
  beforeEach(() => {
    ;(useCommandStore as unknown as jest.Mock).mockImplementation(() => ({
      setOpen: mockSetOpen
    }))
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should render SearchBar correctly', async () => {
    const { container } = renderComponent()
    // Wait for any state updates to complete
    await screen.findByRole('button')
    expect(container).toMatchSnapshot()
  })

  it('should display search icon and text', () => {
    renderComponent()

    expect(screen.getByLabelText('search icon')).toBeInTheDocument()
    expect(screen.getByText('Search')).toBeInTheDocument()
  })

  it('should display keyboard shortcut', () => {
    renderComponent()

    expect(
      screen.getByRole('textbox', { name: 'keyboard shortcut' })
    ).toBeInTheDocument()
    expect(screen.getByText('âŒ˜')).toBeInTheDocument()
    expect(screen.getByText('K')).toBeInTheDocument()
  })

  it('should open command menu when clicked', async () => {
    const user = userEvent.setup()
    renderComponent()

    const button = screen.getByRole('button')
    await user.click(button)

    expect(mockSetOpen).toHaveBeenCalledWith(true)
  })
})
