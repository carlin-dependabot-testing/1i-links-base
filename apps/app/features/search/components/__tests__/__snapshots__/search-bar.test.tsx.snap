// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should render SearchBar correctly 1`] = `
<div>
  <div
    class="group/sidebar-wrapper flex min-h-svh w-full has-[[data-variant=inset]]:bg-sidebar"
    style="--sidebar-width: 16rem; --sidebar-width-icon: 3rem;"
  >
    <script
      nonce=""
    >
      ((e,s,n,m,o,l,u,h)=&gt;{let c=document.documentElement,T=["light","dark"];function p(d){(Array.isArray(e)?e:[e]).forEach(g=&gt;{let k=g==="class",S=k&&l?o.map(f=&gt;l[f]||f):o;k?(c.classList.remove(...S),c.classList.add(d)):c.setAttribute(g,d)}),C(d)}function C(d){h&&T.includes(d)&&(c.style.colorScheme=d)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(m)p(m);else try{let d=localStorage.getItem(s)||n,g=u&&d==="system"?a():d;p(g)}catch(d){}})("class","theme","system",null,["light","dark"],null,true,true)
    </script>
    <div
      class="relative"
    >
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 w-full gap-2"
      >
        <svg
          aria-label="search icon"
          class="lucide lucide-search mr-2 h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <path
            d="m21 21-4.3-4.3"
          />
        </svg>
        <span>
          Search
        </span>
        <kbd
          aria-label="keyboard shortcut"
          class="pointer-events-none ml-auto inline-flex h-5 select-none items-center gap-1 rounded border bg-muted px-1.5 font-mono text-[10px] font-medium opacity-100"
          role="textbox"
        >
          <span
            class="text-xs"
          >
            âŒ˜
          </span>
          K
        </kbd>
      </button>
    </div>
    <div
      class="fixed bottom-4 right-4 z-50 flex items-center gap-2 rounded-lg bg-background/80 p-2 shadow-sm backdrop-blur-sm"
    />
    <div
      aria-label="Notifications (F8)"
      role="region"
      style="pointer-events: none;"
      tabindex="-1"
    >
      <ol
        class="fixed top-0 z-[100] flex max-h-screen w-full flex-col-reverse p-4 sm:bottom-0 sm:right-0 sm:top-auto sm:flex-col md:max-w-[420px]"
        tabindex="-1"
      />
    </div>
  </div>
</div>
`;
